File: backend/app/utils/markdown.py
"""
Shared markdown processing utilities
This implementation mirrors the frontend markdown service to ensure consistency
"""

import re


class MarkdownService:
    """
    Markdown processing service with consistent behavior across frontend/backend.
    """

    H1_REGEX = re.compile(r"^#\s+(.+)$", re.MULTILINE)
    H1_REMOVE_REGEX = re.compile(r"^#\s+(.+)($|\n)", re.MULTILINE)
    DEFAULT_TITLE = "Untitled Note"

    def extract_title(self, content: str) -> str:
        """
        Extract title from markdown content.
        Looks for the first H1 heading (# Title).
        """
        match = self.H1_REGEX.search(content)
        return match.group(1).strip() if match else self.DEFAULT_TITLE

    def format_content(self, content: str) -> str:
        """
        Format note content to ensure proper structure.
        Ensures content has proper spacing and formatting.
        """
        cleaned = content.strip()
        title = self.extract_title(cleaned)

        return f"# {title}\n\n" if cleaned == f"# {title}" else cleaned

    def remove_h1(self, content: str) -> str:
        """
        Remove the H1 title heading from markdown content.
        Used to avoid duplicating the title in rendered content.
        Captures the newline after the heading to avoid extra blank lines.
        Only removes the first H1 heading found.
        """
        return self.H1_REMOVE_REGEX.sub("", content, count=1).strip()

    def create_empty_note(self, title: str = "New Note") -> str:
        """Create a new empty note content template."""
        return f"# {title}\n\n"


# Singleton instance for convenience
markdown_service = MarkdownService()

File: ui/src/utils/markdown.ts
/**
 * Shared markdown processing utilities
 * This implementation mirrors the backend markdown service to ensure consistency
 */

export interface MarkdownProcessor {
  extractTitle(content: string): string;
  formatContent(content: string): string;
  removeH1(content: string): string;
  createEmptyNote(title?: string): string;
}

export class MarkdownService implements MarkdownProcessor {
  private static readonly H1_REGEX = /^#\s+(.+)$/m;
  private static readonly H1_REMOVE_REGEX = /^#\s+(.+)($|\n)/m;
  private static readonly DEFAULT_TITLE = 'Untitled Note';

  extractTitle(content: string): string {
    const match = content.match(MarkdownService.H1_REGEX);
    return match ? match[1].trim() : MarkdownService.DEFAULT_TITLE;
  }

  formatContent(content: string): string {
    const cleaned = content.trim();
    const title = this.extractTitle(cleaned);

    return cleaned === `# ${title}` ? `# ${title}\n\n` : cleaned;
  }

  removeH1(content: string): string {
    // JavaScript replace() without /g flag only replaces the first occurrence
    return content.replace(MarkdownService.H1_REMOVE_REGEX, '').trim();
  }

  createEmptyNote(title: string = 'New Note'): string {
    return `# ${title}\n\n`;
  }
}

// Export singleton instance for convenience
export const markdownService = new MarkdownService();

// Shared test cases to ensure parity with backend
export const markdownTestCases = [
  { input: '# Test\n\nContent', expectedTitle: 'Test' },
  { input: 'No heading', expectedTitle: 'Untitled Note' },
  { input: '## Not H1\n\nContent', expectedTitle: 'Untitled Note' },
  { input: '# Multiple\n\n# Headings', expectedTitle: 'Multiple' },
  { input: '# Title with spaces  ', expectedTitle: 'Title with spaces' },
  { input: '#NoSpace', expectedTitle: 'Untitled Note' },
  { input: '', expectedTitle: 'Untitled Note' },
];
